flytekit/core/type_engine.py

```diff
@@ -502,22 +502,27 @@ def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], exp
 
         self._make_dataclass_serializable(python_val, python_type)
 
-        # The function looks up or creates a JSONEncoder specifically designed for the object's type.
-        # This encoder is then used to convert a data class into a JSON string.
-        try:
-            encoder = self._encoder[python_type]
-        except KeyError:
-            encoder = JSONEncoder(python_type)
-            self._encoder[python_type] = encoder
+        # The `to_json` integrated through mashumaro's `DataClassJSONMixin` allows for more
+        # functionality than JSONEncoder
+        if isinstance(python_val, DataClassJSONMixin):
+            json_str = python_val.to_json()
```

Comment:
Can this check for `to_json` and use duck typing?

```suggestion
        if hasattr(python_val, "to_json"):
            json_str = python_val.to_json()
```

Reply:
For some reason I don't fully understand, this was causing some other tests in the repo to fail locally. The duck-type approach was actually what I started with (and the version we have running in our repo).

Reply:
To be specific, these 7 tests fail consistently if I use duck typing instead of the explicit `DataClassJSONMixin` check:
```
FAILED tests/flytekit/unit/core/test_type_engine.py::test_optional_flytefile_in_dataclass - AssertionError: assert equals failed
FAILED tests/flytekit/unit/core/test_type_engine.py::test_flyte_file_in_dataclass - mashumaro.exceptions.InvalidFieldValue: Field "b" of type test_flyte_file_in_dataclass.<locals>.TestInnerFileStruct in test_flyt...
FAILED tests/flytekit/unit/core/test_type_hints.py::test_flyte_file_in_dataclass - TypeError: Error encountered while executing 'wf':
FAILED tests/flytekit/unit/core/test_type_engine.py::test_flyte_directory_in_dataclass - assert equals failed
FAILED tests/flytekit/unit/core/test_type_hints.py::test_structured_dataset_in_dataclass - AssertionError: assert equals failed
FAILED tests/flytekit/unit/core/test_type_engine.py::test_structured_dataset_in_dataclass - mashumaro.exceptions.InvalidFieldValue: Field "a" of type StructuredDataset in test_structured_dataset_in_dataclass.<locals>.Dat...
FAILED tests/flytekit/unit/core/test_dataclass.py::test_dataclasses_json_mixin_with_flyte_types - TypeError: Error encountered while executing 'nested_dc_wf':
```

Reply:
@thomasjpfan yes this resolves the opened issue as-is, because it's strictly the mashumaro `DataClassJSONMixin` that has different behavior than their JSON encoder.